File: 2025/02 February/14 Fixing Two nodes of a BST/main.cpp
// Link: https://www.geeksforgeeks.org/problems/fixed-two-nodes-of-a-bst/1

#include <iostream>
#include <queue>
#include <sstream>
#include <algorithm>
using namespace std;

class Node
{
public:
    int data;
    Node *left, *right;

    Node(int val)
    {
        data = val;
        left = right = NULL;
    }
};

Node *buildTree(string str)
{
    // Corner Case
    if (str.length() == 0 || str[0] == 'N')
        return NULL;

    // Creating vector of strings from input
    // string after spliting by space
    vector<string> ip;

    istringstream iss(str);
    for (string str; iss >> str;)
        ip.push_back(str);

    // Create the root of the tree
    Node *root = new Node(stoi(ip[0]));

    // Push the root to the queue
    queue<Node *> queue;
    queue.push(root);

    // Starting from the second element
    int i = 1;
    while (!queue.empty() && i < ip.size())
    {

        // Get and remove the front of the queue
        Node *currNode = queue.front();
        queue.pop();

        // Get the current node's value from the string
        string currVal = ip[i];

        // If the left child is not null
        if (currVal != "N")
        {

            // Create the left child for the current node
            currNode->left = new Node(stoi(currVal));

            // Push it to the queue
            queue.push(currNode->left);
        }

        // For the right child
        i++;
        if (i >= ip.size())
            break;
        currVal = ip[i];

        // If the right child is not null
        if (currVal != "N")
        {

            // Create the right child for the current node
            currNode->right = new Node(stoi(currVal));

            // Push it to the queue
            queue.push(currNode->right);
        }
        i++;
    }

    return root;
}

bool isBST(Node *n, int lower, int upper)
{
    if (!n)
        return true;
    if (n->data <= lower || n->data >= upper)
        return false;
    return (isBST(n->left, lower, n->data) && isBST(n->right, n->data, upper));
}

bool compare(Node *a, Node *b, vector<pair<int, int>> &mismatch)
{
    if (!a && !b)
        return true;
    if (!a || !b)
        return false;

    if (a->data != b->data)
        mismatch.push_back(pair<int, int>(a->data, b->data));

    return (compare(a->left, b->left, mismatch) &&
            compare(a->right, b->right, mismatch));
}

class Solution
{
public:
    void make_BST(Node *root, vector<int> &v, int &i)
    {
        if (root == NULL)
        {
            return;
        }

        make_BST(root->left, v, i);
        root->data = v[i++];
        make_BST(root->right, v, i);
    }

    void LNR(Node *root, vector<int> &v)
    {
        if (root == NULL)
        {
            return;
        }

        LNR(root->left, v);
        v.push_back(root->data);
        LNR(root->right, v);
    }

    void correctBST(Node *root)
    {
        vector<int> v;
        LNR(root, v);

        sort(v.begin(), v.end());
        int i = 0;
        make_BST(root, v, i);
    }
};

int main()
{
    Solution objext;
    string s = "10 5 8 2 20";
    Node *root = buildTree(s);
    objext.correctBST(root);

    s = "5 10 20 2 8";
    root = buildTree(s);
    objext.correctBST(root);

    return 0;
}

File: 2025/02 February/14 Fixing Two nodes of a BST/main.py
# Link: https://www.geeksforgeeks.org/problems/fixed-two-nodes-of-a-bst/1

from collections import deque


# Tree Node
class Node:

    def __init__(self, val):
        self.right = None
        self.data = val
        self.left = None


# Function to Build Tree
def buildTree(s):
    # Corner Case
    if len(s) == 0 or s[0] == "N":
        return None

    # Creating list of strings from input
    # string after spliting by space
    ip = list(map(str, s.split()))

    # Create the root of the tree
    root = Node(int(ip[0]))
    size = 0
    q = deque()

    # Push the root to the queue
    q.append(root)
    size = size + 1

    # Starting from the second element
    i = 1
    while size > 0 and i < len(ip):
        # Get and remove the front of the queue
        currNode = q[0]
        q.popleft()
        size = size - 1

        # Get the current node's value from the string
        currVal = ip[i]

        # If the left child is not null
        if currVal != "N":

            # Create the left child for the current node
            currNode.left = Node(int(currVal))

            # Push it to the queue
            q.append(currNode.left)
            size = size + 1
        # For the right child
        i = i + 1
        if i >= len(ip):
            break
        currVal = ip[i]

        # If the right child is not null
        if currVal != "N":

            # Create the right child for the current node
            currNode.right = Node(int(currVal))

            # Push it to the queue
            q.append(currNode.right)
            size = size + 1
        i = i + 1
    return root


def isBST(n, lower, upper):
    if n is None:
        return True
    if n.data <= lower or n.data >= upper:
        return False
    return isBST(n.left, lower, n.data) and isBST(n.right, n.data, upper)


def compare(a, b, mismatch):
    if a is None and b is None:
        return True
    if a is None or b is None:
        return False

    if a.data != b.data:
        mismatch.append((a.data, b.data))

    return compare(a.left, b.left, mismatch) and compare(a.right, b.right, mismatch)


class Solution:
    def correctBST(self, root):
        def inOrder(node):
            nonlocal prev, first, second
            if node is None:
                return
            inOrder(node.left)
            if prev and prev.data > node.data:
                if not first:
                    first = prev
                second = node
            prev = node
            inOrder(node.right)

        prev = None
        first = None
        second = None
        inOrder(root)
        first.data, second.data = second.data, first.data


if __name__ == "__main__":
    obj = Solution()
    s = "10 5 8 2 20"
    root = buildTree(s)
    obj.correctBST(root)

    s = "5 10 20 2 8"
    root = buildTree(s)
    obj.correctBST(root)
